### PR TITLE
[DS][28/n] Create WillBeRequestedCondition

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Creates a WillBeRequestedCondition. Naming is somewhat sketch, but the basic idea is that in the past we've merged together this concept with existing rules.

I.e. in AMP-world, the "skip_on_parent_missing" rule is actually "skip on parent (missing and will not be requested this tick)".

This is obviously a bit confusing, and now that all of our dep conditions are composable, breaking this out into its own separate condition is desirable.

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21640?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#21641** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21641?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21640** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21640?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#21615** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21615?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21613** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21613?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21612** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21612?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21573** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21573?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21546** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21546?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21545** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21545?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21541** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21541?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21540** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21539** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21538** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21538?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21537** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21537?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21536** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21536?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21535** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21535?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21521** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21521?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21520** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21520?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21511** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21511?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#21512](https://github.com/dagster-io/dagster/pull/21512) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21512?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#21510** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21510?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21508** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21508?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21507** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21507?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21505** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21505?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21504** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21504?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21503** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21503?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21502** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21502?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21501** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21501?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21500** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21500?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21499** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21499?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ion-elgreco** [comment]
Could you add a condition that instead of only working on any or all assets, on just an explicit subset?

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/slice_conditions.py [hunk 1]
```diff
@@ -48,6 +48,31 @@ def compute_slice(self, context: SchedulingContext) -> AssetSlice:
         return context.asset_graph_view.compute_in_progress_asset_slice(context.asset_key)
 
 
+@whitelist_for_serdes
+class WillBeRequestedCondition(SliceSchedulingCondition):
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/slice_conditions.py:52]: nit: I think class name should include the word "tick" just like the static constructor and description

**Review by schrockn — CHANGES_REQUESTED** [review]
Just req'ing changes based on class name to make sure you see it since the name will get serialized and it is hard to change.

**Comment by OwenKephart** [comment]
@ion-elgreco 

> Could you add a condition that instead of only working on any or all assets, it applies to an explicit subset?

Adding that functionality to `SchedulingCondition.any_deps_match` and similar is certainly in the plans -- currently conceptualizing adding `allow_keys` and `ignore_keys`  type arguments.

**Review by schrockn — APPROVED** [review]

**Comment by OwenKephart** [comment]
### Merge activity

* **May 14, 10:38 AM EDT**: @OwenKephart started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/21640).

**Pull request merged**